### PR TITLE
Enabling hierarchical multiplatform project structure

### LINKED ISSUE #1006: Enable hierarchical project structure
We should move to proper hierarchical project setup as the new default and we could clean up our source set configuration and take advantage of the new cinterop commonizer to share code across MacOS and iOS, etc.

https://kotlinlang.org/docs/multiplatform-hierarchy.html

### PR DESCRIPTION
**Pull request by rorbech**
This PR rearranges the source sets and enables [Hierarchical Multi Platform Project structure ](https://kotlinlang.org/docs/multiplatform-hierarchy.html)

Closes #1006

### CODE REVIEW

File: packages/cinterop/build.gradle.kts
Comment (cmelchior) [inline packages/cinterop/build.gradle.kts:209]: Aren't this one created automatically? https://kotlinlang.org/docs/multiplatform-share-on-platforms.html#share-code-in-libraries

This would also impact a lot of wiring below in this file.

Update:
After trying to build without creating this, I can see it is failing, which I guess is why you are creating it and after re-reading the documentation it became clear that the above link was just an example, and the only supported shortcuts are these ones 🙈 https://kotlinlang.org/docs/multiplatform-share-on-platforms.html#use-target-shortcuts

File: packages/gradle.properties
Comment (cmelchior) [inline packages/gradle.properties:25]: Should we keep a reference to this YouTrack, since it is still relevant?

File: packages/library-sync/build.gradle.kts
Comment (cmelchior) [inline packages/library-sync/build.gradle.kts:44]:
```suggestion
    macosX64()
```

File: packages/test-base/build.gradle.kts
Comment (cmelchior) [inline packages/test-base/build.gradle.kts:205]: So the emulator is started automatically for us now?

File: packages/test-sync/build.gradle.kts
Comment (cmelchior) [inline packages/test-sync/build.gradle.kts:65]:
```suggestion
                // shouldn't be an issue as it is already a transitive dependency of library-sync.
```

**Review by cmelchior — COMMENTED** [review]
Looking really nice. Still playing around with so just leaving initial feedback. We should also update https://github.com/realm/realm-kotlin/blob/master/CONTRIBUTING.md#multiplatform-source-layout to reflect the new hierachy.

File: packages/test-base/build.gradle.kts
Comment (rorbech) [reply packages/test-base/build.gradle.kts:205]: Yep. Handled by the Kotlin gradle plugin ... the output after removing this is 
```
> Task :test-base:iosX64Test
Starting process 'command '/usr/bin/xcrun''. Working directory: /Users/claus.rorbech/proj/realm-kotlin/packages/test-base Command: /usr/bin/xcrun simctl spawn --standalone iPhone 12 /Users/claus.rorbech/proj/realm-kotlin/packages/test-base/build/bin/iosX64/debugTest/test.kexe -- --ktest_no_exit_code --ktest_logger=TEAMCITY
```

File: packages/gradle.properties
Comment (rorbech) [reply packages/gradle.properties:25]: The issue is specific to composite gradle builds which we don't use anymore. And since we are in fact enabling hierarchical structure it seems a bit weird to keep. If we should keep it, then it would maybe better to place it at the workaround for composite builds not working at https://github.com/realm/realm-kotlin/blob/master/packages/test-base/build.gradle.kts#L23?

File: packages/cinterop/build.gradle.kts
Comment (rorbech) [reply packages/cinterop/build.gradle.kts:209]: 🤪  ... 👍

File: packages/gradle.properties
Comment (rorbech) [reply packages/gradle.properties:25]: I guess it is actually already referenced (or at least the duplicated issue) in https://github.com/realm/realm-kotlin/blob/master/CONTRIBUTING.md#advanced-project-setup which is probably the right place to keep the reference as it covers the overall project setup.

File: packages/cinterop/build.gradle.kts
Comment (nhachicha) [inline packages/cinterop/build.gradle.kts:621]:
```suggestion
        else -> "Unsupported macOS architecture"
```

File: packages/gradle.properties
Comment (nhachicha) [inline packages/gradle.properties:24]: Since we're using Kotlin `1.7.20` on master these flags are not necessary I think (they're  opt-in pre 1.6.20)

File: packages/library-base/build.gradle.kts
Comment (nhachicha) [inline packages/library-base/build.gradle.kts:112]: can't we use `val iosArm64Main by getting`?

File: packages/test-base/build.gradle.kts
Comment (nhachicha) [inline packages/test-base/build.gradle.kts:205]: I suppose we can still set the `iosDevice` property in Gradle properties/Env to request a specific simulator to align local/CI tests

**Review by nhachicha — COMMENTED** [review]
Nice refactor 🙌. minor comments, also I still have some red lines in `nativeDarwinTest`regarding some imports
image
not sure here why `coreErrorAsThrowable` is implemented inside an object for JVM and as global function for darwin?

**Review by cmelchior — APPROVED** [review]
Nice...now I can run all tests from the IDE 🚀 

IMO this looks ready to merge once the merge conflicts are fixed. One thought though. Does it make sense that we symlink folders to native and jvm anymore? Don't we only need the symlink in `androidAndroidTest`?  But it isn't critical and can be addressed later.

File: packages/gradle.properties
Comment (rorbech) [reply packages/gradle.properties:24]: But isn't it better to apply what we need explicitly? 🤔

File: packages/test-base/build.gradle.kts
Comment (rorbech) [reply packages/test-base/build.gradle.kts:205]: Maybe a good option to keep. Hooked the `iosDevice-property into the Gradle plugins infrastructure with https://github.com/realm/realm-kotlin/pull/1085/files#diff-ad7d5c068deb652e9518763651bed6874ac941679d187d260bec712efec3cf27R198

**Comment by rorbech** [comment]
> IMO this looks ready to merge once the merge conflicts are fixed. One thought though. Does it make sense that we symlink folders to native and jvm anymore? Don't we only need the symlink in `androidAndroidTest`? But it isn't critical and can be addressed later.

The reason why we are sym-linking tests are not due to hierarchical builds not working, but because we would otherwise not be able to trigger Android tests from the IDE. Details are in https://github.com/realm/realm-kotlin/blob/master/CONTRIBUTING.md#test-organization, but basically it is because the KMP project for some reason have chosen to [execute Android tests without an Android context](https://youtrack.jetbrains.com/issue/KT-42298/KMM-plugin-Run-common-tests-on-Android-devices#focus=Comments-27-4657970.0-0) instead of just using `jvmTest` for that 🤯 🤷

**Comment by rorbech** [comment]
We could consider to move the actual test files to `commonTest` and change the direction of the symlink so that `androidAndroidTest` links to the `commonTest` ones. This would enable us to have common test where you select platform from the gutter icon dialog, except for `androidTests. This would require explicit removal of `commonTest` source set from the `androidTest` to avoid duplicate symbols a bit like suggested in https://youtrack.jetbrains.com/issue/KT-34712/MPP-Gradle-Plugin-provide-a-way-to-suppress-commonTest-dependsOn#focus=Comments-27-6290261.0-0

**Pull request merged**
